### PR TITLE
[No Ticket] React Native: Background Fetch PoC

### PR DESCRIPTION
**Pull request by briansztamfater**
## A bit of context

Most (or a lot) of React Native applications are built on top of Expo toolchain, a framework that eliminates the needs of leading with Xcode or Android Studio stuff (native modules) and makes development faster. Although Expo is a really awesome and useful framework, at the time of this PR being created, it does not support important mobile features such as Background Fetch, Bluetooth APIs or Background Geolocation which are needed in a lot of projects, if not the most. For supporting features that are not built-in implemented on and not compatible with Expo, the framework provides a way to `eject` from Expo, which will give developers back the control of native modules and use ExpoKit library to continue using Expo APIs, with the big drawback of leading again with Xcode and Android Studio projects (native modules). In that way, devs can use external packages without any inconvenient, but project initial configuration or future SDK / packages upgrades can be a mess.

## About this PoC

The idea of this PoC is showing how to create a React Native project with `create-react-native-app`, eject it from Expo while continue using ExpoKit library, and integrate the following Background Fetching cross-platform library: https://github.com/transistorsoft/react-native-background-fetch
The result would be a simple React Native project with native modules running on iOS and Android, that logs an input every time the background service is executed.

### Requisites

- Have CRNA installed ([How to install CRNA](https://github.com/react-community/create-react-native-app#installation))
- Have Expo CLI installed ([How to install Expo CLI](https://docs.expo.io/versions/latest/guides/exp-cli.html))
- Have Android Studio installed ([How to install Android Studio](https://developer.android.com/studio/install))
- Have Xcode installed (only macOS) ([How to install Xcode](https://itunes.apple.com/us/app/xcode/id497799835))
- Have CocoaPods installed (only macOS) ([How to install CocoaPods](https://guides.cocoapods.org/using/getting-started.html#installation))

#### Create project

- On Terminal, run `$ create-react-native-app <ProjectName>`
- Then go to your project's directory `$ cd <ProjectName>`

#### Ejecting

- Check your Expo version on package.json and app.json and update to the latest if necessary
- `$ yarn run eject`
- Follow further instructions
- Add `expokit` package if it's not there on `package.json`

#### Running our app

```
$ npm install
$ exp start
```

##### iOS (macOS only)

```
$ cd ios
$ pod install
```
- Open `.xcworkspace` file on Xcode and select `Build & Run`
- Now you can start writing Javascript code!

##### Android

- Open `settings.gradle` file on Android Studio and let it configure the project
- Build and run on an emulator or Android device
- Now you can start writing Javascript code!

### Integrating [react-native-background-fetch](https://github.com/transistorsoft/react-native-background-fetch)

```
$ npm install react-native-background-fetch --save
$ react-native link
```

#### iOS

- Follow [CocoaPods installation process](https://github.com/transistorsoft/react-native-background-fetch/blob/master/docs/INSTALL-COCOAPODS-IOS.md)
- Build & Run

#### Android

- Open `build.gradle` file and add `maven { url "$rootDir/../node_modules/react-native-background-fetch/android/libs" }`
- Sync gradle
- Build & Run

#### React Native

Finally, some JavaScript code!

Open App.js and replace the whole file with:
```
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import BackgroundFetch from "react-native-background-fetch";

export default class App extends React.Component {

  componentDidMount() {
    // Configure it.
    BackgroundFetch.configure({
      minimumFetchInterval: 15, // <-- minutes (15 is minimum allowed)
      stopOnTerminate: false,   // <-- Android-only,
      startOnBoot: true         // <-- Android-only
    }, () => {
      console.log("Received background-fetch event");
      // Required: Signal completion of your task to native code
      // If you fail to do this, the OS can terminate your app
      // or assign battery-blame for consuming too much background-time
      BackgroundFetch.finish(BackgroundFetch.FETCH_RESULT_NEW_DATA);
    }, (error) => {
      console.log("RNBackgroundFetch failed to start");
    });

    // Optional: Query the authorization status.
    BackgroundFetch.status((status) => {
      switch(status) {
        case BackgroundFetch.STATUS_RESTRICTED:
          console.log("BackgroundFetch restricted");
          break;
        case BackgroundFetch.STATUS_DENIED:
          console.log("BackgroundFetch denied");
          break;
        case BackgroundFetch.STATUS_AVAILABLE:
          console.log("BackgroundFetch is enabled");
          break;
      }
    });
  }

  render() {
    return (
      <View style={styles.container}>
        <Text>Background Fetch Proof of Concept!</Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

### Debugging

- See [Debugging section from react-native-background-fetch repository](https://github.com/transistorsoft/react-native-background-fetch#debugging)
- You can see the logs on the terminal session where you ran `exp start` previously

## Reviewers required

@mvazquez-ms @mmaldini33 @mravinale 

Feel free to comment any suggestion or question you have 😄

### CODE REVIEW

**Review by mravinale — APPROVED** [review]
Nice work!!

File: ReactNative/BackgroundTask/android/.gradle/buildOutputCleanup/cache.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+#Fri Sep 07 01:57:59 ART 2018
```
Comment (mvazquez-ms) [inline ReactNative/BackgroundTask/android/.gradle/buildOutputCleanup/cache.properties:1]: Could these `.grade` files be ignored  (`.gitignore`) or are they required for some reason? ([here](https://www.gitignore.io/api/gradle)'s a typical Gradle gitignore)

File: ReactNative/BackgroundTask/ios/backgroundtask/AppDelegate.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+
+#import <UIKit/UIKit.h>
+
+@interface AppDelegate : UIResponder <UIApplicationDelegate>
```
Comment (mvazquez-ms) [inline ReactNative/BackgroundTask/ios/backgroundtask/AppDelegate.h:5]: This has to be written in Obj-C? or could it be done in Swift as well?

**Review by mvazquez-ms — APPROVED** [review]
Awesome work! love the description on the PR.

Just left some two minor comments, and I _think_ there's a couple of things missing in the `gitignore` (e.g. `.DS_Store` or `/xcuserdata/briansztamfater.xcuserdatad/UserInterfaceState.xcuserstate`)

**Pull request closed without merging**